numpy/core/src/multiarray/datetime_strings.c

```diff
@@ -323,22 +326,37 @@ parse_iso_8601_datetime(char const *str, Py_ssize_t len,
     if (len == 3 && tolower(str[0]) == 'n' &&
                     tolower(str[1]) == 'o' &&
                     tolower(str[2]) == 'w') {
-        NPY_TIME_T rawtime = 0;
         PyArray_DatetimeMetaData meta;
 
-        time(&rawtime);
+        // initialize the _PyTime_t struct
+        _PyTime_t rawtime = _PyTime_GetSystemClock();
+        _Py_clock_info_t rawtime_info = {" ", 0, 1, 0.0};
+
+        // actual rawtime we use is from here,
+        // along with resolution info
+        _PyTime_GetSystemClockWithInfo(&rawtime, &rawtime_info);
+        double resolution = rawtime_info.resolution;
+
+        // set resolutions to 1/1000 observed
+        // floating point comparisons need to be cleaner
+        if ((resolution < 1.01) && (resolution > 0.0011))  {
+            bestunit = NPY_FR_ms;
+        }
+        else if ((resolution < 1) && (resolution > 0.0009)) {
+            bestunit = NPY_FR_us;
+        }
+        else if ((resolution < 0.001) && (resolution > 0.0000009)) {
+            bestunit = NPY_FR_ns;
+        }
+        else if (resolution < 0.0000000011) {
+            bestunit = NPY_FR_ps;
+        }
```

Comment:
This block feels like a terrible hack--I'm not sure what span of resolution units we reasonably need to check for, but surely there is a cleaner way to do the float comparisons, etc.